packages/enclave-contracts/contracts/lib/ExitQueueLib.sol

```diff
@@ -0,0 +1,340 @@
+        uint256 ticketAmount,
+        uint256 licenseAmount,
+        bool includedLockedAssets
+    );
+
+    error ZeroAmountNotAllowed();
+    error TimestampOverflow();
+    error IndexOutOfBounds();
+
+    function queueAssetsForExit(
+        ExitQueueState storage state,
+        address operator,
+        uint64 exitDelaySeconds,
+        uint256 ticketAmount,
+        uint256 licenseAmount
+    ) internal {
+        if (ticketAmount == 0 && licenseAmount == 0) {
+            return;
+        }
+
+        uint64 currentTimestamp = uint64(block.timestamp);
+        require(
+            currentTimestamp <= (type(uint64).max - exitDelaySeconds),
+            TimestampOverflow()
+        );
+        uint64 unlockTimestamp = currentTimestamp + exitDelaySeconds;
+
+        ExitTranche[] storage operatorQueue = state.operatorQueues[operator];
+
+        uint256 len = operatorQueue.length;
+        bool merged;
+        if (len != 0) {
+            ExitTranche storage lastTranche = operatorQueue[len - 1];
+            if (lastTranche.unlockTimestamp == unlockTimestamp) {
+                if (ticketAmount != 0) lastTranche.ticketAmount += ticketAmount;
+                if (licenseAmount != 0)
+                    lastTranche.licenseAmount += licenseAmount;
+                merged = true;
+            }
+        }
+
+        if (!merged) {
+            operatorQueue.push();
+            ExitTranche storage t = operatorQueue[len];
+            t.unlockTimestamp = unlockTimestamp;
+            t.ticketAmount = ticketAmount;
+            t.licenseAmount = licenseAmount;
+        }
+
+        _updatePendingTotals(
+            state,
+            operator,
+            ticketAmount,
+            licenseAmount,
+            true
+        );
+
+        emit AssetsQueuedForExit(
+            operator,
+            ticketAmount,
+            licenseAmount,
+            unlockTimestamp
+        );
+    }
+
+    function queueTicketsForExit(
+        ExitQueueState storage state,
+        address operator,
+        uint64 exitDelaySeconds,
+        uint256 ticketAmount
+    ) internal {
+        queueAssetsForExit(state, operator, exitDelaySeconds, ticketAmount, 0);
+    }
+
+    function queueLicensesForExit(
+        ExitQueueState storage state,
+        address operator,
+        uint64 exitDelaySeconds,
+        uint256 licenseAmount
+    ) internal {
+        queueAssetsForExit(state, operator, exitDelaySeconds, 0, licenseAmount);
+    }
+
+    function getPendingAmounts(
+        ExitQueueState storage state,
+        address operator
+    ) internal view returns (uint256 ticketAmount, uint256 licenseAmount) {
+        PendingAmounts storage pending = state.pendingTotals[operator];
+        return (pending.ticketAmount, pending.licenseAmount);
+    }
+
+    function previewClaimableAmounts(
+        ExitQueueState storage state,
+        address operator
+    ) internal view returns (uint256 ticketAmount, uint256 licenseAmount) {
```

Comment:
not really familiar with writing libraries, assume it's ok to have view here because it's not modifying its own state?

Reply:
The function doesn't actually modify any state, just reads state, so it can be marked as view.

Reply:
yea right it modifies `operatorQueue ` which is  part of that `state` storage variable - wasn't really sure how it'd work with that actual being a storage update on some other contract (not own ofc as libraries do not hold any state)

Reply:
iiuc, `operatorQueue` is a storage pointer. The storage slot that it points to is not modified. Similar for `tranche`.